### PR TITLE
Fixed the (do/dr) error occuring in windows while performing the tests

### PR DESCRIPTION
**Pull request by Pranavchiku**

### CODE REVIEW

**Comment by certik** [comment]
Thanks!

I think this is mixing the old PR with the new one. Why don't you send a PR just for the `run_tests.py` fixes? That might work and we can merge it.

File: src/libasr/compiler_tester/tester.py [hunk 1]
```diff
@@ -121,6 +121,19 @@ def fixdir(s: bytes) -> bytes:
...
+    b = bytearray(b)
+    i = 0
+    while i + 1 <= len(b):
+        if b[i] == nl_cr:
+            if b[i + 1] == nl_lf:
+                del b[i]
+            else:
+                b[i] = nl_lf
+        i += 1
+    return bytes(b)
```
[the remark this answers is not in the document]
Comment (Pranavchiku) [reply src/libasr/compiler_tester/tester.py:136]: It returned the following error
```
Traceback (most recent call last):
  File "C:\Users\lenovo\lfortran\run_tests.py", line 219, in <module>
    tester_main("LFortran", single_test)
  File "C:\Users\lenovo\lfortran\src/libasr\compiler_tester\tester.py", line 386, in tester_main
    single_test(test,
  File "C:\Users\lenovo\lfortran\run_tests.py", line 45, in single_test
    run_test(
  File "C:\Users\lenovo\lfortran\src/libasr\compiler_tester\tester.py", line 295, in run_test
    jo = run(basename, cmd, os.path.join("tests", "output"), infile=infile,
  File "C:\Users\lenovo\lfortran\src/libasr\compiler_tester\tester.py", line 195, in run
    temp=unl_loop_del(open(infile, "rb").read())
  File "C:\Users\lenovo\lfortran\src/libasr\compiler_tester\tester.py", line 137, in unl_loop_del
    return b.replace("\r\n", "\n")
TypeError: a bytes-like object is required, not 'str'
```

**Comment by Pranavchiku** [comment]
> Thanks!
> 
> I think this is mixing the old PR with the new one. Why don't you send a PR just for the `run_tests.py` fixes? That might work and we can merge it.

Yeah sure! before merging I am facing another error for subroutine2.f90 trying to resolve that, pointed that on zulip

File: src/libasr/compiler_tester/tester.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (konradha) [reply src/libasr/compiler_tester/tester.py:136]: It's a bytes-like object, so (to be verbose)

```python
b.replace(bytes('\r\n', encoding='utf-8'), bytes('\n', encoding='utf-8'))
```

should do the trick.

[the remark this answers is not in the document]
Comment (Pranavchiku) [reply src/libasr/compiler_tester/tester.py:136]: Verifying it in few minutes!

**Comment by czgdp1807** [comment]
So the history of this PR is messed up because in one of your previous PRs you directly edited the main branch and then pushed into your fork. Here's how to fix this,

First some assumptions,

1. You have two branches, `main` (the lead branch) and `integration_test` (this PR's branch).
2. Assuming you are in `integration_test` branch locally.
3. The name of your fork's remote is say, `your_remote`.
4. Assuming you wanted to apply only 47df3c32d53e24e5b and 180c421567ea5c31.

Steps,

1. Set the remote URL of `origin` correctly. Run, `git remote set-url origin https://github.com/lfortran/lfortran`. If this doesn't work, then only, run `git remote add origin https://github.com/lfortran/lfortran`.
2. Verify the output of `git remote get-url your_remote` is same as `https://github.com/Pranavchiku/lfortran`.
3. `git checkout main` (see 2nd assumption above).
5. `git checkout -b main_backup`
6. `git branch -D main`
7. `git fetch origin main`
8. `git checkout main`
9. `git checkout integration_test`
10. `git checkout -b integration_test_backup`
11. `git branch -D integration_test`
12. `git checkout main`
13. `git checkout -b integration_test`
14. `git cherry-pick 47df3c32d53e24e5b`. If it errors, then only, `git cherry-pick --abort` and manually apply the changes from https://github.com/lfortran/lfortran/pull/121/commits/47df3c32d53e24e5b240d05a75f2b5f1f5309e96. Then `git commit -m "nice message". ` By manually I mean copy and paste changes. If cherry-pick succeeds then move onto next step.
15. `git cherry-pick 180c421567ea5c31`. If errors then `git cherry-pick --abort` and manually apply changes from https://github.com/lfortran/lfortran/pull/121/commits/180c421567ea5c3159f663b0aa144ed2eec29ea0. Then `git commit -m "nice message". If cherry-pick succeeds then move onto next step.
16. `git push -f your_remote integration_test`. This PR will be updated automatically. If you think the PR diff is okay for your then ping us.

**Comment by certik** [comment]
@Pranavchiku go ahead and do what @czgdp1807 wrote. If you run into any issue please ask.

**Comment by Pranavchiku** [comment]
Sure, I will complete that by midnight!

**Comment by Pranavchiku** [comment]
Done, that, hopefully not messed up this time :)

**Comment by czgdp1807** [comment]
The git history looks clean.

File: src/libasr/compiler_tester/tester.py [hunk 2]
```diff
@@ -171,39 +176,41 @@ def run(basename: str, cmd: Union[pathlib.Path, str],
...
+        temp=unl_loop_del(open(stderr_file, "rb").read())
+        stderr_hash = hashlib.sha224(temp).hexdigest()
         stderr_file = os.path.basename(stderr_file)
     else:
         stderr_hash = None
     data = {
         "basename": basename,
         "cmd": cmd,
-        "infile": infile,
+        "infile": infile.replace("\\\\","\\").replace("\\","/"),
```
Comment (certik) [inline src/libasr/compiler_tester/tester.py:213]: Is this change needed? It seems this is already applied on the line 166.

Comment (Pranavchiku) [reply src/libasr/compiler_tester/tester.py:213]: no, it's repeated, changed that.

Comment (certik) [reply src/libasr/compiler_tester/tester.py:213]: Are you saying it is not needed? If so, can you please push a commit to remove this change?

Comment (Pranavchiku) [reply src/libasr/compiler_tester/tester.py]: It is not needed, I removed that and pushed :)

File: src/libasr/compiler_tester/tester.py [hunk 3]
```diff
@@ -171,32 +176,34 @@ def run(basename: str, cmd: Union[pathlib.Path, str],
     if len(r.stdout):
         stdout_file = os.path.join(out_dir, basename + "." + "stdout")
         open(stdout_file, "wb").write(fixdir(r.stdout))
+
```
Comment (certik) [inline src/libasr/compiler_tester/tester.py:179]: Did you put the empty line here for a reason?

If it was just a leftover from coding, then I would remove it, it seems it looks better without it:
```suggestion
```

File: src/libasr/compiler_tester/tester.py [hunk 4]
```diff
@@ -290,6 +301,7 @@ def run_test(testname, basename, cmd, infile, update_reference=False,
 
     dr = json.load(open(jr))
     if do != dr:
+
```
Comment (certik) [inline src/libasr/compiler_tester/tester.py:304]:
```suggestion
```

File: src/libasr/compiler_tester/tester.py [hunk 5]
```diff
@@ -215,6 +222,7 @@ def run(basename: str, cmd: Union[pathlib.Path, str],
     }
     json_file = os.path.join(out_dir, basename + "." + "json")
     json.dump(data, open(json_file, "w"), indent=4)
+    # print("json file is: " + json_file)
```
Comment (certik) [inline src/libasr/compiler_tester/tester.py:225]:
```suggestion
```

File: src/libasr/compiler_tester/tester.py [hunk 6]
```diff
@@ -272,8 +280,11 @@ def run_test(testname, basename, cmd, infile, update_reference=False,
     s = f"{testname} * {basename}"
     basename = bname(basename, cmd, infile)
     infile = os.path.join("tests", infile)
+
+
```
Comment (certik) [inline src/libasr/compiler_tester/tester.py:284]: Probably one empty line is enough:
```suggestion
```
